File: DSAUni/C++/Graphs/provinces.py
def DFS(vis,node,adj, n):
    vis[node] = 1
    
    for i in adj.get(node, []):

        if (not vis[i]):
            DFS(vis,i,adj,n)



def numberofProvinces(adj, n):
    vis = [0] * (n+1)
    count = 0
    for i in range(1,n+1):
        if not vis[i]:
            count+=1
            DFS(vis, i, adj, n)
    return count
def main():
    adj = {
        1: [2],
        2:[1,3],
        3: [2]
    }

    n = 3
    vis = [0] * (n+1)
    provinces = numberofProvinces(adj,n)
    # Time complexity explanation:
    # The queue pops exactly N times (one per node).
    # The inner loop processes all edges across the whole traversal → total 2E for undirected.
    # So total complexity = O(N + 2E) = O(N + E).

    print(provinces)
if __name__ == "__main__":
    main()

File: DSAUni/C++/Graphs/provinces.cpp
#include <bits/stdc++.h>
using namespace std;

void dfs(int node, vector<int> &vis, vector<int> adj[]) {
    vis[node] = 1;
    cout << node << " "; // optional visual trace

    for (auto it : adj[node]) {
        if (!vis[it]) {
            dfs(it, vis, adj);
        }
    }
}

int numberofProvinces(vector<int> &vis, vector<int> adj[], int n) {
    int count = 0;
    for(int i = 1; i <= n; i++) {
        if (!vis[i]) {
            count++;
            dfs(i, vis, adj);
            cout << endl; // separate each province visually
        }
    }
    return count;
}

int main() {
    vector<int> adj[] = {
        {},          // 0 unused
        {2, 3},     // 1
        {5, 6},     // 2
        {1, 4, 7},  // 3
        {3, 8},     // 4
        {2},        // 5
        {2},        // 6
        {3, 8},     // 7
        {4, 7}      // 8
    };

    int n = 8;
    vector<int> vis(n+1, 0);

    cout << "\nDFS traversal by province:\n";
    int provinces = numberofProvinces(vis, adj, n);

    cout << "\nNumber of Provinces: " << provinces << endl;

    return 0;
}
